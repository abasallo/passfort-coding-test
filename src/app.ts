import * as dotenv from 'dotenv'

import express from 'express'
import cors from 'cors'
import bodyParser from 'body-parser'

import index from './express/routes'
import documents from './express/routes/documents'

import { errorMiddleware } from './express/errors'

dotenv.config()

export const app = express()

app.use(cors())

app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.use('/', index)
app.use('/documents', documents)

app.use(errorMiddleware)

export default app
