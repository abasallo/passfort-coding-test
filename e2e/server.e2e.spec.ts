import axios from 'axios'
import http from 'http'
import { ReasonPhrases, StatusCodes } from 'http-status-codes'
import app from '../src/app'

const TEST_SERVER_PORT = 4001

let server: http.Server
beforeAll(async () => {
  server = app.listen({ port: TEST_SERVER_PORT }, () => console.log(`Test Server initialised on port: ${TEST_SERVER_PORT}`))
})

afterAll(() => {
  server.close()
})

describe('Server E2E tests', () => {
  describe('Health Check E2E tests', () => {
    test('Expected response for Health Check', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })

  describe('Documents E2E tests', () => {
    test('Find All Documents', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/documents`,
        data: {
          userId: 3
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Find Document by title', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/documents/title-1`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Find Review by title and timestamp', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/documents/title-2/2024-01-01T00:00:00`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Latest Review by title', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/documents/title-3/latest`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('Add Review', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/documents/title-3`,
        data: {
          content: 'New Content'
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        id: expect.any(Number),
        timestamp: expect.any(String)
      })
    })
  })
})
