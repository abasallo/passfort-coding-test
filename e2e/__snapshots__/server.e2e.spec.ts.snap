// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server E2E tests Documents E2E tests Add Review 1`] = `
{
  "content": "New Content",
  "documentId": 3,
  "id": Any<Number>,
  "timestamp": Any<String>,
}
`;

exports[`Server E2E tests Documents E2E tests Find All Documents 1`] = `
[
  {
    "id": 1,
    "title": "title-1",
  },
  {
    "id": 2,
    "title": "title-2",
  },
  {
    "id": 3,
    "title": "title-3",
  },
]
`;

exports[`Server E2E tests Documents E2E tests Find Document by title 1`] = `
[
  {
    "content": "text-1",
    "documentId": 1,
    "id": 1,
    "timestamp": "2023-01-01T00:00:00",
  },
]
`;

exports[`Server E2E tests Documents E2E tests Find Review by title and timestamp 1`] = `
{
  "content": "text-2-update-1",
  "documentId": 2,
  "id": 3,
  "timestamp": "2023-01-01T00:00:00",
}
`;

exports[`Server E2E tests Documents E2E tests Latest Review by title 1`] = `
{
  "content": "text-3-update-2",
  "documentId": 3,
  "id": 6,
  "timestamp": "2023-01-01T00:00:00",
}
`;

exports[`Server E2E tests Health Check E2E tests Expected response for Health Check 1`] = `"OK"`;
